File: go/pkg/chatgear/conn_pipe.go
package chatgear

import (
	"iter"
	"sync"
	"time"

	"github.com/haivivi/giztoy/go/pkg/audio/codec/opus"
)

// NewPipe creates a connected pair of server and client connections using channels.
// This is useful for testing and in-process communication.
func NewPipe() (*PipeServerConn, *PipeClientConn) {
	// Uplink channels (client -> server)
	uplinkOpus := make(chan StampedOpusFrame, 1024)
	uplinkStates := make(chan *StateEvent, 32)
	uplinkStats := make(chan *StatsEvent, 32)

	// Downlink channels (server -> client)
	downlinkOpus := make(chan StampedOpusFrame, 1024)
	downlinkCmds := make(chan *CommandEvent, 32)

	// Shared error state for cross-connection error propagation
	shared := &pipeSharedState{}

	server := &PipeServerConn{
		uplinkOpus:   uplinkOpus,
		uplinkStates: uplinkStates,
		uplinkStats:  uplinkStats,
		downlinkOpus: downlinkOpus,
		downlinkCmds: downlinkCmds,
		shared:       shared,
	}

	client := &PipeClientConn{
		uplinkOpus:   uplinkOpus,
		uplinkStates: uplinkStates,
		uplinkStats:  uplinkStats,
		downlinkOpus: downlinkOpus,
		downlinkCmds: downlinkCmds,
		shared:       shared,
	}

	return server, client
}

// pipeSharedState holds shared state between server and client connections.
type pipeSharedState struct {
	mu        sync.Mutex
	serverErr error // Error set by server, readable by client
	clientErr error // Error set by client, readable by server
}

// =============================================================================
// PipeServerConn: Server side of pipe connection
// Implements UplinkRx (receive from client) and DownlinkTx (send to client)
// =============================================================================

type PipeServerConn struct {
	// Uplink channels (receive from client)
	uplinkOpus   chan StampedOpusFrame
	uplinkStates chan *StateEvent
	uplinkStats  chan *StatsEvent

	// Downlink channels (send to client)
	downlinkOpus chan StampedOpusFrame
	downlinkCmds chan *CommandEvent

	shared *pipeSharedState

	mu          sync.Mutex
	latestStats *StatsEvent
	closed      bool
}

// --- UplinkRx implementation (receive from client) ---

func (c *PipeServerConn) OpusFrames() iter.Seq2[StampedOpusFrame, error] {
	return func(yield func(StampedOpusFrame, error) bool) {
		for frame := range c.uplinkOpus {
			if !yield(frame, nil) {
				return
			}
		}
		// Get client error from shared state
		c.shared.mu.Lock()
		err := c.shared.clientErr
		c.shared.mu.Unlock()
		if err != nil {
			yield(StampedOpusFrame{}, err)
		}
	}
}

func (c *PipeServerConn) States() iter.Seq2[*StateEvent, error] {
	return func(yield func(*StateEvent, error) bool) {
		for state := range c.uplinkStates {
			if !yield(state, nil) {
				return
			}
		}
		c.shared.mu.Lock()
		err := c.shared.clientErr
		c.shared.mu.Unlock()
		if err != nil {
			yield(nil, err)
		}
	}
}

func (c *PipeServerConn) Stats() iter.Seq2[*StatsEvent, error] {
	return func(yield func(*StatsEvent, error) bool) {
		for stats := range c.uplinkStats {
			c.mu.Lock()
			c.latestStats = stats
			c.mu.Unlock()
			if !yield(stats, nil) {
				return
			}
		}
		c.shared.mu.Lock()
		err := c.shared.clientErr
		c.shared.mu.Unlock()
		if err != nil {
			yield(nil, err)
		}
	}
}

func (c *PipeServerConn) LatestStats() *StatsEvent {
	c.mu.Lock()
	defer c.mu.Unlock()
	return c.latestStats
}

// --- DownlinkTx implementation (send to client) ---

func (c *PipeServerConn) SendOpusFrame(timestamp time.Time, frame opus.Frame) error {
	c.mu.Lock()
	closed := c.closed
	c.mu.Unlock()
	if closed {
		return nil
	}

	select {
	case c.downlinkOpus <- StampedOpusFrame{Timestamp: timestamp, Frame: frame}:
		return nil
	default:
		// Channel full, drop frame
		return nil
	}
}

func (c *PipeServerConn) IssueCommand(cmd Command, t time.Time) error {
	c.mu.Lock()
	closed := c.closed
	c.mu.Unlock()
	if closed {
		return nil
	}

	evt := NewCommandEvent(cmd, t)
	select {
	case c.downlinkCmds <- evt:
		return nil
	default:
		// Channel full, drop command
		return nil
	}
}

// --- Lifecycle ---

func (c *PipeServerConn) Close() error {
	return c.CloseWithError(nil)
}

func (c *PipeServerConn) CloseWithError(err error) error {
	c.mu.Lock()
	defer c.mu.Unlock()
	if c.closed {
		return nil
	}
	c.closed = true

	// Set server error in shared state for client to read
	c.shared.mu.Lock()
	c.shared.serverErr = err
	c.shared.mu.Unlock()

	// Close downlink channels (server owns these)
	close(c.downlinkOpus)
	close(c.downlinkCmds)
	return nil
}

// =============================================================================
// PipeClientConn: Client side of pipe connection
// Implements UplinkTx (send to server) and DownlinkRx (receive from server)
// =============================================================================

type PipeClientConn struct {
	// Uplink channels (send to server)
	uplinkOpus   chan StampedOpusFrame
	uplinkStates chan *StateEvent
	uplinkStats  chan *StatsEvent

	// Downlink channels (receive from server)
	downlinkOpus chan StampedOpusFrame
	downlinkCmds chan *CommandEvent

	shared *pipeSharedState

	mu     sync.Mutex
	closed bool
}

// --- UplinkTx implementation (send to server) ---

func (c *PipeClientConn) SendOpusFrame(timestamp time.Time, frame opus.Frame) error {
	c.mu.Lock()
	closed := c.closed
	c.mu.Unlock()
	if closed {
		return nil
	}

	select {
	case c.uplinkOpus <- StampedOpusFrame{Timestamp: timestamp, Frame: frame}:
		return nil
	default:
		// Channel full, drop frame
		return nil
	}
}

func (c *PipeClientConn) SendState(state *StateEvent) error {
	c.mu.Lock()
	closed := c.closed
	c.mu.Unlock()
	if closed {
		return nil
	}

	select {
	case c.uplinkStates <- state:
		return nil
	default:
		// Channel full, drop state
		return nil
	}
}

func (c *PipeClientConn) SendStats(stats *StatsEvent) error {
	c.mu.Lock()
	closed := c.closed
	c.mu.Unlock()
	if closed {
		return nil
	}

	select {
	case c.uplinkStats <- stats:
		return nil
	default:
		// Channel full, drop stats
		return nil
	}
}

// --- DownlinkRx implementation (receive from server) ---

func (c *PipeClientConn) OpusFrames() iter.Seq2[StampedOpusFrame, error] {
	return func(yield func(StampedOpusFrame, error) bool) {
		for frame := range c.downlinkOpus {
			if !yield(frame, nil) {
				return
			}
		}
		// Get server error from shared state
		c.shared.mu.Lock()
		err := c.shared.serverErr
		c.shared.mu.Unlock()
		if err != nil {
			yield(StampedOpusFrame{}, err)
		}
	}
}

func (c *PipeClientConn) Commands() iter.Seq2[*CommandEvent, error] {
	return func(yield func(*CommandEvent, error) bool) {
		for cmd := range c.downlinkCmds {
			if !yield(cmd, nil) {
				return
			}
		}
		c.shared.mu.Lock()
		err := c.shared.serverErr
		c.shared.mu.Unlock()
		if err != nil {
			yield(nil, err)
		}
	}
}

// --- Lifecycle ---

func (c *PipeClientConn) Close() error {
	return c.CloseWithError(nil)
}

func (c *PipeClientConn) CloseWithError(err error) error {
	c.mu.Lock()
	defer c.mu.Unlock()
	if c.closed {
		return nil
	}
	c.closed = true

	// Set client error in shared state for server to read
	c.shared.mu.Lock()
	c.shared.clientErr = err
	c.shared.mu.Unlock()

	// Close uplink channels (client owns these)
	close(c.uplinkOpus)
	close(c.uplinkStates)
	close(c.uplinkStats)
	return nil
}

// Compile-time interface assertions
var (
	_ UplinkRx   = (*PipeServerConn)(nil)
	_ DownlinkTx = (*PipeServerConn)(nil)
	_ UplinkTx   = (*PipeClientConn)(nil)
	_ DownlinkRx = (*PipeClientConn)(nil)
)

File: rust/chatgear/src/conn_pipe.rs
//! In-memory pipe connection for testing.
//!
//! This module provides a bidirectional pipe for testing client-server communication
//! without actual network connections.

use crate::{
    ConnError, DownlinkRx, DownlinkTx, GearStateEvent, GearStatsEvent, SessionCommandEvent,
    StampedOpusFrame, UplinkRx, UplinkTx,
};
use async_trait::async_trait;
use std::sync::Arc;
use tokio::sync::{mpsc, Mutex, RwLock};

/// Creates a connected pair of server and client connections using channels.
/// This is useful for testing and in-process communication.
pub fn new_pipe() -> (PipeServerConn, PipeClientConn) {
    // Uplink channels (client -> server)
    let (uplink_opus_tx, uplink_opus_rx) = mpsc::channel::<StampedOpusFrame>(1024);
    let (uplink_states_tx, uplink_states_rx) = mpsc::channel(32);
    let (uplink_stats_tx, uplink_stats_rx) = mpsc::channel(32);

    // Downlink channels (server -> client)
    let (downlink_opus_tx, downlink_opus_rx) = mpsc::channel::<StampedOpusFrame>(1024);
    let (downlink_cmds_tx, downlink_cmds_rx) = mpsc::channel(32);

    // Shared error state
    let shared = Arc::new(RwLock::new(PipeSharedState::default()));

    let server = PipeServerConn {
        uplink_opus: Arc::new(Mutex::new(uplink_opus_rx)),
        uplink_states: Arc::new(Mutex::new(uplink_states_rx)),
        uplink_stats: Arc::new(Mutex::new(uplink_stats_rx)),
        downlink_opus: downlink_opus_tx,
        downlink_cmds: downlink_cmds_tx,
        shared: shared.clone(),
        latest_stats: Arc::new(Mutex::new(None)),
        closed: Arc::new(Mutex::new(false)),
    };

    let client = PipeClientConn {
        uplink_opus: uplink_opus_tx,
        uplink_states: uplink_states_tx,
        uplink_stats: uplink_stats_tx,
        downlink_opus: Arc::new(Mutex::new(downlink_opus_rx)),
        downlink_cmds: Arc::new(Mutex::new(downlink_cmds_rx)),
        shared,
        closed: Arc::new(Mutex::new(false)),
    };

    (server, client)
}

/// Shared state between server and client connections.
#[derive(Default)]
struct PipeSharedState {
    server_err: Option<String>,
    client_err: Option<String>,
}

/// Server side of a pipe connection.
/// Implements UplinkRx (receive from client) and DownlinkTx (send to client).
pub struct PipeServerConn {
    // Uplink channels (receive from client)
    uplink_opus: Arc<Mutex<mpsc::Receiver<StampedOpusFrame>>>,
    uplink_states: Arc<Mutex<mpsc::Receiver<GearStateEvent>>>,
    uplink_stats: Arc<Mutex<mpsc::Receiver<GearStatsEvent>>>,

    // Downlink channels (send to client)
    downlink_opus: mpsc::Sender<StampedOpusFrame>,
    downlink_cmds: mpsc::Sender<SessionCommandEvent>,

    shared: Arc<RwLock<PipeSharedState>>,
    latest_stats: Arc<Mutex<Option<GearStatsEvent>>>,
    closed: Arc<Mutex<bool>>,
}

impl PipeServerConn {
    /// Returns the uplink opus receiver for direct access.
    pub fn uplink_opus(&self) -> Arc<Mutex<mpsc::Receiver<StampedOpusFrame>>> {
        self.uplink_opus.clone()
    }

    /// Returns the uplink states receiver for direct access.
    pub fn uplink_states(&self) -> Arc<Mutex<mpsc::Receiver<GearStateEvent>>> {
        self.uplink_states.clone()
    }

    /// Returns the uplink stats receiver for direct access.
    pub fn uplink_stats(&self) -> Arc<Mutex<mpsc::Receiver<GearStatsEvent>>> {
        self.uplink_stats.clone()
    }

    /// Closes the connection with an optional error.
    pub async fn close_with_error(&self, err: Option<String>) -> Result<(), ConnError> {
        let mut closed = self.closed.lock().await;
        if *closed {
            return Ok(());
        }
        *closed = true;

        // Set server error in shared state
        let mut shared = self.shared.write().await;
        shared.server_err = err;

        Ok(())
    }
}

#[async_trait]
impl UplinkRx for PipeServerConn {
    async fn recv_opus_frame(&self) -> Result<Option<StampedOpusFrame>, ConnError> {
        let mut rx = self.uplink_opus.lock().await;
        match rx.recv().await {
            Some(frame) => Ok(Some(frame)),
            None => {
                let shared = self.shared.read().await;
                if let Some(ref err) = shared.client_err {
                    Err(ConnError::ReceiveFailed(err.clone()))
                } else {
                    Ok(None)
                }
            }
        }
    }

    async fn recv_state(&self) -> Result<Option<GearStateEvent>, ConnError> {
        let mut rx = self.uplink_states.lock().await;
        match rx.recv().await {
            Some(state) => Ok(Some(state)),
            None => {
                let shared = self.shared.read().await;
                if let Some(ref err) = shared.client_err {
                    Err(ConnError::ReceiveFailed(err.clone()))
                } else {
                    Ok(None)
                }
            }
        }
    }

    async fn recv_stats(&self) -> Result<Option<GearStatsEvent>, ConnError> {
        let mut rx = self.uplink_stats.lock().await;
        match rx.recv().await {
            Some(stats) => {
                *self.latest_stats.lock().await = Some(stats.clone());
                Ok(Some(stats))
            }
            None => {
                let shared = self.shared.read().await;
                if let Some(ref err) = shared.client_err {
                    Err(ConnError::ReceiveFailed(err.clone()))
                } else {
                    Ok(None)
                }
            }
        }
    }

    async fn latest_stats(&self) -> Option<GearStatsEvent> {
        self.latest_stats.lock().await.clone()
    }

    async fn close(&self) -> Result<(), ConnError> {
        self.close_with_error(None).await
    }
}

#[async_trait]
impl DownlinkTx for PipeServerConn {
    async fn send_opus_frame(&self, frame: &StampedOpusFrame) -> Result<(), ConnError> {
        self.downlink_opus
            .send(frame.clone())
            .await
            .map_err(|e| ConnError::SendFailed(e.to_string()))
    }

    async fn send_command(&self, cmd: &SessionCommandEvent) -> Result<(), ConnError> {
        self.downlink_cmds
            .send(cmd.clone())
            .await
            .map_err(|e| ConnError::SendFailed(e.to_string()))
    }

    async fn close(&self) -> Result<(), ConnError> {
        self.close_with_error(None).await
    }
}

/// Client side of a pipe connection.
/// Implements UplinkTx (send to server) and DownlinkRx (receive from server).
pub struct PipeClientConn {
    // Uplink channels (send to server)
    uplink_opus: mpsc::Sender<StampedOpusFrame>,
    uplink_states: mpsc::Sender<GearStateEvent>,
    uplink_stats: mpsc::Sender<GearStatsEvent>,

    // Downlink channels (receive from server)
    downlink_opus: Arc<Mutex<mpsc::Receiver<StampedOpusFrame>>>,
    downlink_cmds: Arc<Mutex<mpsc::Receiver<SessionCommandEvent>>>,

    shared: Arc<RwLock<PipeSharedState>>,
    closed: Arc<Mutex<bool>>,
}

impl PipeClientConn {
    /// Returns the downlink opus receiver for direct access.
    pub fn downlink_opus(&self) -> Arc<Mutex<mpsc::Receiver<StampedOpusFrame>>> {
        self.downlink_opus.clone()
    }

    /// Returns the downlink commands receiver for direct access.
    pub fn downlink_cmds(&self) -> Arc<Mutex<mpsc::Receiver<SessionCommandEvent>>> {
        self.downlink_cmds.clone()
    }

    /// Closes the connection with an optional error.
    pub async fn close_with_error(&self, err: Option<String>) -> Result<(), ConnError> {
        let mut closed = self.closed.lock().await;
        if *closed {
            return Ok(());
        }
        *closed = true;

        // Set client error in shared state
        let mut shared = self.shared.write().await;
        shared.client_err = err;

        Ok(())
    }
}

#[async_trait]
impl UplinkTx for PipeClientConn {
    async fn send_opus_frame(&self, frame: &StampedOpusFrame) -> Result<(), ConnError> {
        self.uplink_opus
            .send(frame.clone())
            .await
            .map_err(|e| ConnError::SendFailed(e.to_string()))
    }

    async fn send_state(&self, state: &GearStateEvent) -> Result<(), ConnError> {
        self.uplink_states
            .send(state.clone())
            .await
            .map_err(|e| ConnError::SendFailed(e.to_string()))
    }

    async fn send_stats(&self, stats: &GearStatsEvent) -> Result<(), ConnError> {
        self.uplink_stats
            .send(stats.clone())
            .await
            .map_err(|e| ConnError::SendFailed(e.to_string()))
    }

    async fn close(&self) -> Result<(), ConnError> {
        self.close_with_error(None).await
    }
}

#[async_trait]
impl DownlinkRx for PipeClientConn {
    async fn recv_opus_frame(&self) -> Result<Option<StampedOpusFrame>, ConnError> {
        let mut rx = self.downlink_opus.lock().await;
        match rx.recv().await {
            Some(frame) => Ok(Some(frame)),
            None => {
                let shared = self.shared.read().await;
                if let Some(ref err) = shared.server_err {
                    Err(ConnError::ReceiveFailed(err.clone()))
                } else {
                    Ok(None)
                }
            }
        }
    }

    async fn recv_command(&self) -> Result<Option<SessionCommandEvent>, ConnError> {
        let mut rx = self.downlink_cmds.lock().await;
        match rx.recv().await {
            Some(cmd) => Ok(Some(cmd)),
            None => {
                let shared = self.shared.read().await;
                if let Some(ref err) = shared.server_err {
                    Err(ConnError::ReceiveFailed(err.clone()))
                } else {
                    Ok(None)
                }
            }
        }
    }

    async fn close(&self) -> Result<(), ConnError> {
        self.close_with_error(None).await
    }
}

#[cfg(test)]
mod pipe_tests {
    use super::*;
    use crate::{GearState, SetVolume, UplinkRx, UplinkTx, DownlinkRx, DownlinkTx};

    #[tokio::test]
    async fn test_pipe_creation() {
        let (server, client) = new_pipe();
        // Just test that they can be created
        drop(server);
        drop(client);
    }

    #[tokio::test]
    async fn test_pipe_opus_client_to_server() {
        let (server, client) = new_pipe();

        // Client sends stamped opus frame
        let frame = StampedOpusFrame::now(vec![0xFC, 0x01, 0x02, 0x03]);
        client.send_opus_frame(&frame).await.unwrap();

        // Server receives it
        let received = server.recv_opus_frame().await.unwrap().unwrap();
        assert_eq!(received.frame, frame.frame);
    }

    #[tokio::test]
    async fn test_pipe_opus_server_to_client() {
        let (server, client) = new_pipe();

        // Server sends stamped opus frame
        let frame = StampedOpusFrame::now(vec![0xFC, 0x04, 0x05, 0x06]);
        server.send_opus_frame(&frame).await.unwrap();

        // Client receives it
        let received = client.recv_opus_frame().await.unwrap().unwrap();
        assert_eq!(received.frame, frame.frame);
    }

    #[tokio::test]
    async fn test_pipe_state_event() {
        let (server, client) = new_pipe();

        // Client sends state
        let state = GearStateEvent::new(GearState::Recording);
        client.send_state(&state).await.unwrap();

        // Server receives it
        let received = server.recv_state().await.unwrap().unwrap();
        assert_eq!(received.state, GearState::Recording);
    }

    #[tokio::test]
    async fn test_pipe_stats_event() {
        let (server, client) = new_pipe();

        // Client sends stats
        let mut stats = GearStatsEvent::new();
        stats.volume = Some(crate::Volume {
            percentage: 75.0,
            update_at: giztoy_jsontime::Milli::now(),
        });
        client.send_stats(&stats).await.unwrap();

        // Server receives it
        let received = server.recv_stats().await.unwrap().unwrap();
        assert_eq!(received.volume.unwrap().percentage, 75.0);
    }

    #[tokio::test]
    async fn test_pipe_command() {
        let (server, client) = new_pipe();

        // Server sends command
        let cmd = SetVolume::new(50);
        let event = SessionCommandEvent::new(&cmd);
        server.send_command(&event).await.unwrap();

        // Client receives it
        let received = client.recv_command().await.unwrap().unwrap();
        assert_eq!(received.cmd_type, "set_volume");
    }

    #[tokio::test]
    async fn test_pipe_close() {
        let (server, client) = new_pipe();

        // Close client
        client.close_with_error(None).await.unwrap();

        // Close server
        server.close_with_error(None).await.unwrap();

        // Both should be marked as closed
        // Note: In Rust, recv() only returns None when all senders are dropped,
        // which doesn't happen until the struct is dropped.
    }

    #[tokio::test]
    async fn test_pipe_close_with_error() {
        let (_server, client) = new_pipe();

        // Close client with error
        client
            .close_with_error(Some("test error".to_string()))
            .await
            .unwrap();

        // Verify double close is idempotent
        client.close_with_error(None).await.unwrap();
    }

    #[tokio::test]
    async fn test_pipe_bidirectional() {
        let (server, client) = new_pipe();

        // Simultaneous communication
        let client_frame = StampedOpusFrame::now(vec![0x01, 0x02]);
        let server_frame = StampedOpusFrame::now(vec![0x03, 0x04]);

        client.send_opus_frame(&client_frame).await.unwrap();
        server.send_opus_frame(&server_frame).await.unwrap();

        let from_client = server.recv_opus_frame().await.unwrap().unwrap();
        let from_server = client.recv_opus_frame().await.unwrap().unwrap();

        assert_eq!(from_client.frame, client_frame.frame);
        assert_eq!(from_server.frame, server_frame.frame);
    }
}
